File: Compiler/src/main/kotlin/nodes/ExprSeqNode.java
package nodes;

public class ExprSeqNode {
    public ExprNode first;
    public ExprNode last;

    public ExprSeqNode(ExprNode expr) {
        this.first = expr;
        this.last = expr;
    }

    public void addExpr(ExprNode expr) {
        this.last.next = expr;
        this.last = expr;
    }

    public static ExprSeqNode createTableConstructorExprSeqNode(FieldListNode tableConstructor) {
        ExprNode exprNodeTable = new ExprNode();
        exprNodeTable.id = Utils.getLastId();
        exprNodeTable.type = ExprType.TABLE_CONSTRUCTOR;
        exprNodeTable.table_constructor = tableConstructor;
        return new ExprSeqNode(exprNodeTable);
    }

    public static ExprSeqNode createStringExprSeqNode(String string) {
        ExprNode exprNodeString = new ExprNode();
        exprNodeString.id = Utils.getLastId();
        exprNodeString.type = ExprType.STRING;
        exprNodeString.string_value = string;
        return new ExprSeqNode(exprNodeString);
    }

    public static ExprSeqNode createVarListNode(VarNode var) {
        ExprNode expr_var_node = new ExprNode();
        expr_var_node.id = Utils.getLastId();
        expr_var_node.type = ExprType.VAR;
        expr_var_node.var = var;
        return new ExprSeqNode(expr_var_node);
    }

    public void addVar(VarNode var) {
        ExprNode expr_var_node = new ExprNode();
        expr_var_node.id = Utils.getLastId();
        expr_var_node.type = ExprType.VAR;
        expr_var_node.var = var;
        this.last.next = expr_var_node;
        this.last = expr_var_node;
    }
}

File: Compiler/src/main/kotlin/nodes/ExprSeqNode.kt
package nodes

import org.checkerframework.common.returnsreceiver.qual.This
import javax.swing.TransferHandler

class ExprSeqNode(var first: ExprNode) {
    var last: ExprNode = first

    fun addExpr(expr: ExprNode) {
        last.next = expr
        last = expr
    }

    fun length() : Int {
        var result = 0
        var current: ExprNode? = this.first
        while (current != null) {
            result++
            current = current.next
        }
        return result
    }

    fun addVar(varNode: VarNode?) {
        val exprVarNode = ExprNode()
        exprVarNode.id = Utils.lastId
        exprVarNode.type = ExprType.VAR
        exprVarNode.varNode = varNode
        last.next = exprVarNode
        last = exprVarNode
    }

    companion object {
        fun createTableConstructorExprSeqNode(tableConstructor: FieldListNode?): ExprSeqNode {
            val exprNodeTable = ExprNode()
            exprNodeTable.id = Utils.lastId
            exprNodeTable.type = ExprType.TABLE_CONSTRUCTOR
            exprNodeTable.tableConstructor = tableConstructor
            return ExprSeqNode(exprNodeTable)
        }

        fun createStringExprSeqNode(string: String?): ExprSeqNode {
            val exprNodeString = ExprNode()
            exprNodeString.id = Utils.lastId
            exprNodeString.type = ExprType.STRING
            exprNodeString.stringValue = string!!
            return ExprSeqNode(exprNodeString)
        }

        fun createVarListNode(varNode: VarNode?): ExprSeqNode {
            val exprVarNode = ExprNode()
            exprVarNode.id = Utils.lastId
            exprVarNode.type = ExprType.VAR
            exprVarNode.varNode = varNode
            return ExprSeqNode(exprVarNode)
        }
    }
}